File: Simulation/498. Diagonal Traverse.cpp
class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& mat) 
    {
        map<int, vector<int>> mp;
        int n = mat.size(), m = mat[0].size();
        for(int i = 0; i < n; i++)
        {
            for(int j = 0; j < m; j++)
            {
                int idx = i + j;
                if(idx % 2) mp[idx].push_back(mat[i][j]);
                else mp[idx].insert(mp[idx].begin(),mat[i][j]);
            }
        }

        vector<int> ans;
        for(auto x : mp)
            for(int v:x.second) ans.push_back(v);
        
        return ans;
    }
};

File: Simulation/498. Diagonal Traverse.py
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        store = defaultdict(deque)
        n, m = len(mat), len(mat[0])
        for i in range(n):
            for j in range(m):
                index_sum = i + j
                if index_sum % 2:
                    store[index_sum].append(mat[i][j])
                else:
                    store[index_sum].appendleft(mat[i][j])
        return [val for index_sum in store for val in store[index_sum]]